extmod/moduselect.c

```diff
@@ -275,7 +275,7 @@ STATIC mp_obj_t poll_poll(uint n_args, const mp_obj_t *args) {
             ret_list->items[n_ready++] = mp_obj_new_tuple(2, tuple);
             if (self->flags & FLAG_ONESHOT) {
                 // Don't poll next time, until new event flags will be set explicitly
```

Comment:
> But I want to think on it a bit more. It changes the semantics of one-shot behaviour. Eg what if you register an object for read and write polling at the same time in one-shot mode, then perhaps you'd expect it to only inform you once about being ready for either reading or writing.

Yes, that can be seen even from this comment in the code. The original specification for one-shot mode is simple: it fires once, and doesn't fire again until reinstated explicitly.

Reply:
In addition to that, one-shot feature doesn't come on its own, but was inspired by the EPOLLONESHOT flag from Linux' epoll() (uasyncio was initially implemented using epoll; while the underlying mechanism was switched to poll() for wider POSIX compatibility, being able to "accelerate" it on Linux is definitely remains a feature point). Only reviewing Linux source could answer all details of how that flag works, but human-language description corresponds to how FLAG_ONESHOT was implemented in the uselect module:

~~~
       EPOLLONESHOT (since Linux 2.6.2)
              Sets the one-shot behavior for the associated  file  descriptor.
              This  means that after an event is pulled out with epoll_wait(2)
              the associated file descriptor is  internally  disabled  and  no
              other  events will be reported by the epoll interface.  The user
              must call epoll_ctl()  with  EPOLL_CTL_MOD  to  rearm  the  file
              descriptor with a new event mask.
~~~